pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/SomaticResultUploadService.java

```diff
@@ -0,0 +1,186 @@
+package org.broadinstitute.dsm.service;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+import com.google.auth.ServiceAccountSigner;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.HttpMethod;
+import com.typesafe.config.Config;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.broadinstitute.ddp.client.GoogleBucketClient;
+import org.broadinstitute.ddp.constants.ConfigFile;
+import org.broadinstitute.ddp.exception.DDPException;
+import org.broadinstitute.ddp.util.ConfigUtil;
+import org.broadinstitute.ddp.util.GoogleCredentialUtil;
+import org.broadinstitute.dsm.db.SomaticResultUpload;
+import org.broadinstitute.dsm.model.somatic.result.SomaticResultMetaData;
+import org.broadinstitute.dsm.model.somatic.result.SomaticResultUploadSettings;
+import org.broadinstitute.dsm.statics.ApplicationConfigConstants;
+import org.owasp.fileio.FileValidator;
+
+@Slf4j
+public class SomaticResultUploadService {
+    private final ServiceAccountSigner signer;
+    private final GoogleBucketClient storageClient;
+    private final Map<String, String> realmToUploadBucketMap;
+    private final int maxSignedUrlMins;
+
+    private final SomaticResultUploadSettings somaticUploadSettings;
+
+    public static SomaticResultUploadService fromConfig(Config cfg) {
+        String signerJson = ConfigUtil.toJson(cfg.getConfig(ApplicationConfigConstants.FILE_DOWNLOAD_CREDENTIALS));
+        InputStream signerStream = new ByteArrayInputStream(signerJson.getBytes(StandardCharsets.UTF_8));
+        ServiceAccountCredentials signerCredentials;
+        try {
+            signerCredentials = ServiceAccountCredentials.fromStream(signerStream);
+        } catch (IOException e) {
+            throw new DDPException("Could not get signer credentials", e);
+        }
+
+        GoogleCredentials bucketCredentials;
+        boolean ensureDefault = false;
+        bucketCredentials = GoogleCredentialUtil.initCredentials(ensureDefault);
+        if (bucketCredentials == null) {
+            log.error("Could not get bucket credentials, defaulting to signer credentials");
+            bucketCredentials = signerCredentials;
+        }
+
+        String projectId = cfg.getString(ApplicationConfigConstants.FILE_DOWNLOAD_PROJECT_ID);
```

Comment:
Just curious, what does 'project' refer to?

Reply:
The GCP `project` name.